### ISSUE TITLE
Recommend using `exec()` with `await` for better stack traces

### ISSUE THREAD

**Issue by vkarpov15** [Automattic/mongoose#8747]
Re: #8691

**Comment by vkarpov15** [comment]
I dug into this and unfortunately there isn't much we can do other than capture a stack trace on every query function call. The fundamental issue is that `await` on a custom thenable executes `then()` as a separate microtask, for example:

```javascript
'use strict';
  
class Q {
  exec() { 
    const err = new Error('test');
    return new Promise((resolve, reject) => {
      setTimeout(() => reject(err), 10);
    });
  }

  then(resolve, reject) {
    return this.exec().then(resolve, reject);
  }
}

run().catch(err => console.log(err));

async function run() {
  const q = new Q();
  await q;
}
```

Prints:

```
$ ~/Workspace/libs/node-v10.17.0-linux-x64/bin/node gh-8747.js 
Error: test
    at Q.exec (/home/val/Workspace/MongoDB/troubleshoot-mongoose/gh-8747.js:5:17)
    at Q.then (/home/val/Workspace/MongoDB/troubleshoot-mongoose/gh-8747.js:12:17)
    at process._tickCallback (internal/process/next_tick.js:68:7)
    at Function.Module.runMain (internal/modules/cjs/loader.js:834:11)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)
```

In Node.js 12 the stack trace is unhelpful even with [async stack traces](https://thecodebarbarian.com/async-stack-traces-in-node-js-12):

```
$ ~/Workspace/libs/node-v12.13.1-linux-x64/bin/node --async-stack-traces gh-8747.js 
Error: test
    at Q.exec (/home/val/Workspace/MongoDB/troubleshoot-mongoose/gh-8747.js:5:17)
    at Q.then (/home/val/Workspace/MongoDB/troubleshoot-mongoose/gh-8747.js:12:17)
```

TLDR; we should update our query docs to recommend using `exec()` with `await` for better stack traces.

**Comment by wootwoot1234** [comment]
There has to be a workaround for this, right?

**Comment by vkarpov15** [comment]
@wootwoot1234 yes, use `exec()` :+1:

**Comment by alko89** [comment]
What about adding a config that will set `exec()` behavior as default?

**Comment by vkarpov15** [comment]
@alko89 unfortunately that isn't possible AFAIK because of how JavaScript handles custom [thenables](https://masteringjs.io/tutorials/fundamentals/thenable). We already use [exec under the hood to allow `await query`](https://github.com/Automattic/mongoose/blob/71478c4ff64a8afa6007c9d246e4403f701ed48a/lib/query.js#L4982-L4984), but that still doesn't help.
